### PR TITLE
Move executor pod construction to a separate class.

### PR DESCRIPTION
**Pull request by mccheah**
This is the first of several measures to make KubernetesClusterSchedulerBackend feasible to test. Requires #445 but only for convenience and not semantically speaking.

The idea is to start breaking down the functionality of KubernetesClusterSchedulerBackend into multiple individually unit-test friendly units. The logic that builds the executor pod structure was by far the single longest method that could be isolated with relative ease.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+          .addNewVolumeMount()
+          .withName(FilenameUtils.getBaseName(dir))
+          .withMountPath(dir)
+          .endVolumeMount()
+          .build()
+      }
+    }.getOrElse(containerWithExecutorLimitCores)
+    val withMaybeShuffleConfigPod = shuffleServiceConfig.map { config =>
+      config.shuffleDirs.foldLeft(executorPod) { (builder, dir) =>
+        new PodBuilder(builder)
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala:241]: Ah we lost the indentation here, I'll fix that.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+      ConfigurationUtils.combinePrefixedKeyValuePairsWithDeprecatedConf(
+          sparkConf,
+          KUBERNETES_EXECUTOR_ANNOTATION_PREFIX,
+          KUBERNETES_EXECUTOR_ANNOTATIONS,
+          "executor annotation")
+  private val nodeSelector =
+      ConfigurationUtils.parsePrefixedKeyValuePairs(
+          sparkConf,
+          KUBERNETES_NODE_SELECTOR_PREFIX,
+          "node-selector")
```
Comment (foxish) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: "node selector" for consistency?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala [hunk 3]
```diff
@@ -234,18 +234,28 @@ private[spark] class KubernetesSuite extends SparkFunSuite with BeforeAndAfter {
...
+    val executorJvmOptionsFile = storeJvmOptionsInTempFile(
+        Map("simpleExecutorConf" -> "simpleExecutorConfValue",
+            "executor conf with spaces" -> "executor conf with spaces value"),
+        "executor-jvm-options.properties",
+        "JVM options that should be set on the executors.")
     sparkConf.set(SparkLauncher.DRIVER_EXTRA_JAVA_OPTIONS,
         "-DsimpleDriverConf=simpleDriverConfValue" +
             " -Ddriverconfwithspaces='driver conf with spaces value'")
-    sparkConf.set("spark.files", driverJvmOptionsFile.getAbsolutePath)
+    sparkConf.set(SparkLauncher.EXECUTOR_EXTRA_JAVA_OPTIONS,
```
Comment (foxish) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala]: This is a separate change from the refactor correct?

**Review by foxish — COMMENTED** [review]
Refactor LGTM

**Comment by mccheah** [comment]
I believe the diff is corrupted and I have to fix it in git.

**Comment by mccheah** [comment]
Actually I just need to rebase against branch-2.2-kubernetes and make the diff that way.

File: resource-managers/kubernetes/README.md [hunk 4]
```diff
@@ -14,14 +14,11 @@ important matters to keep in mind when developing this feature.
...
-To build Spark with Kubernetes support, use the `kubernetes` profile when invoking Maven. For example, to simply compile the Kubernetes core implementation module along with its dependencies:
+To build Spark with Kubernetes support, use the `kubernetes` profile when invoking Maven. For example, to simply compile
+the Kubernetes core implementation module along with its dependencies:
 
     build/mvn compile -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
 
-If this is the first time you compile the Kubernetes core implementation module, run the following command to install the dependencies and compile:
-
-    build/mvn install -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
-
```
Comment (ash211) [inline resource-managers/kubernetes/README.md]: please don't revert this change -- it went in in a PR and somehow your PRs keep coming close to reverting it.. ?

File: resource-managers/kubernetes/README.md [hunk 5]
```diff
@@ -14,14 +14,11 @@ important matters to keep in mind when developing this feature.
 
 # Building Spark with Kubernetes Support
 
-To build Spark with Kubernetes support, use the `kubernetes` profile when invoking Maven. For example, to simply compile the Kubernetes core implementation module along with its dependencies:
+To build Spark with Kubernetes support, use the `kubernetes` profile when invoking Maven. For example, to simply compile
+the Kubernetes core implementation module along with its dependencies:
 
     build/mvn compile -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
```
Comment (mccheah) [inline resource-managers/kubernetes/README.md]: Believe this change is unintentional

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala (hunk 2 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: That would be a break in the configuration I think. Aside from that, `SparkConf` keys have never had spaces in them.

**Comment by mccheah** [comment]
@foxish rebase complete.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+import scala.collection.JavaConverters._
+
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.deploy.kubernetes.{ConfigurationUtils, InitContainerResourceStagingServerSecretPlugin, PodWithDetachedInitContainer, SparkPodInitContainerBootstrap}
+import org.apache.spark.deploy.kubernetes.config._
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.deploy.kubernetes.submit.{InitContainerUtil, MountSmallFilesBootstrap}
+import org.apache.spark.util.Utils
+
+// Strictly an extension of KubernetesClusterSchedulerBakcne that is factored out for testing.
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: typo: `KubernetesClusterSchedulerBakcne`

Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: by extension, you mean the method in this trait is the same as a method in `KubernetesClusterSchedulerBackend` ?  That implies to me it should do multiple inheritance.

I'm not sure the "strictly an extension" language makes sense -- maybe instead say that it's only used in the scheduler backend?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala (hunk 2 above)
Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: I think that last string is only used for log output -- `node selector` seems like it would be fine

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    sparkConf: SparkConf,
+    nodeAffinityExecutorPodModifier: NodeAffinityExecutorPodModifier,
+    mountSmallFilesBootstrap: Option[MountSmallFilesBootstrap],
+    executorInitContainerBootstrap: Option[SparkPodInitContainerBootstrap],
+    executorMountInitContainerSecretPlugin: Option[InitContainerResourceStagingServerSecretPlugin])
+  extends ExecutorPodFactory {
+
+  import ExecutorPodFactoryImpl._
+
+  private val EXECUTOR_ID_COUNTER = new AtomicLong(0L)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: where is this used?  I think it should be deleted because KubernetesClusterSchedulerBackend.scala already has one

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/NodeAffinityExecutorPodModifier.scala [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+package org.apache.spark.scheduler.cluster.kubernetes
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import io.fabric8.kubernetes.api.model.{Pod, PodBuilder}
+
+import org.apache.spark.deploy.kubernetes.constants.ANNOTATION_EXECUTOR_NODE_AFFINITY
+import org.apache.spark.internal.Logging
+
+// Strictly an extension of ExecutorPodFactory but extracted out for testing.
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/NodeAffinityExecutorPodModifier.scala]: not sure this comment adds much -- it's good practice to have smaller more modular pieces anyway for understanding, regardless of testing purposes

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala (hunk 1 above)
Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala:241]: looks fixed now

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala (hunk 6 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala]: Extension meaning a plugin, or functionality that is pretty much only used in the scheduler backend.

**Comment by ash211** [comment]
@mccheah conflicts on the GiB -> MiB conversion fix since you moved that elsewhere

**Comment by mccheah** [comment]
rerun integration tests please

**Comment by ash211** [comment]
Rerun unit tests please

**Comment by mccheah** [comment]
@aash @foxish good to merge this?

**Comment by ash211** [comment]
test coverage is unchanged (this is an internal refactor) and enables more granular testing in followup PRs
